pkg/storaged/client.js

```diff
@@ -206,6 +214,203 @@ function is_multipath_master(block) {
+                    const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                    if (!subvolid_match && !subvol_match) {
+                        console.warn("findmnt entry without subvol and subvolid", fs);
+                        break;
+                    }
+
+                    const { subvolid } = subvolid_match.groups;
+                    const { subvol } = subvol_match.groups;
+                    const subvolume = {
+                        pathname: subvol,
+                        id: subvolid,
+                        mount_points: [fs.target],
+                    };
+
+                    if (!(fs.uuid in btrfs_mounts)) {
+                        btrfs_mounts[fs.uuid] = { };
+                    }
+
+                    // We need to handle multiple mounts, they are listed seperate.
+                    if (subvolid in btrfs_mounts[fs.uuid]) {
+                        btrfs_mounts[fs.uuid][subvolid].mount_points.push(fs.target);
+                    } else {
+                        btrfs_mounts[fs.uuid][subvolid] = subvolume;
+                    }
+                }
+            }
+        } catch (exc) {
+            if (exc.message)
+                console.error("unable to parse findmnt JSON output", exc);
+        }
+
+        // Update client state
+        if (!deep_equal(client.btrfs_mounts, btrfs_mounts)) {
+            client.btrfs_mounts = btrfs_mounts;
+            debug("btrfs_findmnt_poll mounts:", client.btrfs_mounts);
+            client.update();
+        }
+    };
+
+    const findmnt_poll = () => {
+        return cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "--poll"], { superuser: true, err: "message" }).stream(() => {
+            cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,TARGET", "--json"],
+                          { superuser: true, err: "message" }).then(output => update_btrfs_mounts(output)).catch(err => {
+                // When there are no btrfs filesystems left this can fail and thus we need to manually reset the mount info.
+                client.btrfs_mounts = {};
+                client.update();
+                if (err.message) {
+                    console.error("findmnt exited with an error", err);
+                }
+            });
+        });
+    };
+
+    // This fails when no btrfs filesystem is found with the --mtab option and exits with 1, so that is kinda useless, however without --mtab
+    // we don't get a nice flat structure. So we ignore the errors
+    cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,SOURCE,TARGET", "--json"],
+                  { superuser: true, err: "message" }).then(output => {
+        update_btrfs_mounts(output);
+        findmnt_poll();
+    }).catch(err => {
+        // only log error when there is a real issue.
+        if (client.superuser.allowed && err.message) {
+            console.error(`unable to run findmnt ${err}`);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20240111-164332-85e5434d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)